fuel-relayer/src/pending_events.rs

```diff
@@ -0,0 +1,375 @@
+            self.bundled_removed_eth_events
+                .push((eth_block, vec![event]));
+        }
+    }
+
+    /// Handle eth log events
+    /// Done
+    pub async fn handle_eth_log(&mut self, event: EthEventLog, eth_block: u64, removed: bool) {
+        if removed {
+            self.bundle_removed_events(event, eth_block);
+            return;
+        }
+        // apply all reverted event
+        if !self.bundled_removed_eth_events.is_empty() {
+            info!(target:"relayer", "Reorg happened on ethereum. Reverting {} logs",self.bundled_removed_eth_events.len());
+
+            // if there is new log that is not removed it means we can revert our pending removed eth events.
+            let mut current_da_height = 0;
+            for (da_height, _) in std::mem::take(&mut self.bundled_removed_eth_events).into_iter() {
+                if da_height != current_da_height {
+                    self.pending.pop_back();
+                    current_da_height = da_height;
+                }
+            }
+        }
+        // apply new event to pending queue
+        self.append_eth_events(&event, eth_block).await;
+    }
+
+    /// At begining we will ignore all event until event for new fuel block commit commes
+    /// after that syncronization can start.
+    /// Done
+    pub async fn append_eth_events(&mut self, fuel_event: &EthEventLog, da_height: u64) {
+        if let Some(front) = self.pending.back() {
+            if front.da_height != da_height {
+                self.pending.push_back(PendingDiff::new(da_height))
+            }
+        } else {
+            self.pending.push_back(PendingDiff::new(da_height))
+        }
+        let last_diff = self.pending.back_mut().unwrap();
+        match *fuel_event {
+            EthEventLog::AssetDeposit {
+                account,
+                token,
+                amount,
+                deposit_nonce,
+                ..
+            } => {
+                last_diff
+                    .assets_deposited
+                    .insert(deposit_nonce, (account, token, amount));
+            }
+            EthEventLog::ValidatorDeposit { depositor, deposit } => {
+                // overflow is not possible
+                *last_diff.stake_diff.entry(depositor).or_insert(0) += deposit as i64;
+            }
+            EthEventLog::ValidatorWithdrawal {
+                withdrawer,
+                withdrawal,
+            } => {
+                // underflow should not be possible and it should be restrained by contract
+                *last_diff.stake_diff.entry(withdrawer).or_insert(0) -= withdrawal as i64;
+            }
+            EthEventLog::FuelBlockCommited { .. } => {
+                // TODO do nothing? or maybe update some state as BlockCommitSeen, BlockCommitFinalized, etc.
+            }
+        }
+    }
+
+    /// Used in two places. On initial sync and when new fuel blocks is
+    /// Done
+    pub async fn flush_validator_diffs(
```

Comment:
this affects more than just the validator diffs, it seems like this is actually handling all the different kinds of diffs.

```suggestion
    pub async fn commit_diffs(
```